### PR TITLE
issue NewValidation: CSCwi17652 check for service-ep flag

### LINKED ISSUE #294: NewValidation: CSCwi17652 check for service-ep flag
1. Version check between   5.2.5c/6.0.1g and 16.0.8e/6.1.1f. 

### What needs to be validated
2. config validation:

This is related to the fix made under https://cdetsng.cisco.com/summary/#/defect/CSCwi17652

Between 5.2.5c/6.0.1g and 16.0.8e/6.1.1f as part of the fix tied to a different bug CSCwb17734, the Service-ep flag is set on the service epg (vlanCktEp) even when PBR is not enabled. 

		moquery -c vlan.CktEp -f 'vlan.CktEp.encap=="vlan-65"'
		# vlan.CktEp
		encap                    : vlan-65
		adminSt                  : active
		allowUsegUnsupported     : 0
		childAction              :
		classPrefOperSt          : encap
		ctrl                     : policy-enforced,service-ep
		name                     : GZ-Test-DC-Tenant:DC-LBctxGZ-Test-DC-VRF:LB_Outside:

When customer upgrades to a version  >= 16.0.8e/6.1.1f, Service-ep flag gets removed for the specific service epg (vlanCktEp).

		SV-40G-Leaf211# moquery -c vlan.CktEp -f 'vlan.CktEp.encap=="vlan-93"'
		Total Objects shown: 1

		# vlan.CktEp
		encap                    : vlan-93
		adminSt                  : active
		allowUsegUnsupported     : 0
		childAction              :
		classPrefOperSt          : encap
		ctrl                     : policy-enforced
		epgDn                    : uni/tn-CIB/LDevInst-[uni/tn-CIB/lDevVip-Prod-SV-LB]-ctx-DC-VRF/G-Prod-SV-LBctxDC-VRF-N-Prod-LB-Out_BD-C-Prod-SV-LB-Out

This may affect working service graphs which uses the service epg (vlanCktEp) and has valid pbr config.

The R comments for CSCwi17652 has the fix details which resulted in the removal of the service-ep flag.
	Problem:
	Service-ep flag is set on the vlanCktEp of service epg when PBR is not enabled. 
	This leads to DL bit being set on traffic coming from service device, and hence more BUM traffic on customer network.

	RCA:
	- This is due to regression caused by fix of “CSCwb17734 Lafayeete 16.0.0.76 has traffic failure 30 % drop in s1-leaf2 due to SECURITY_GROUP_DENY” in 5.2.
	- https://aci-gerrit.cisco.com:8443/c/mgmt/+/246518
		- The commit sets service-ep knob on vlanCktEp during
	EthIfBI::configureIfConn()—>vlan::CktEpBI::configure()
		- But we are not checking if EPpRedirPol (PBR) is present. This leads to service-ep being set even when there is no PBR.

	Fix:
	- In CktEpBI::postExplicitCb(), we are handling service-ep setting during vlanCktEp create, modify and delete. 
	So ideally we need not set service-ep flags during vlanCktEp's factory. So removed fix for CSCwb17734.

### Why it needs to be validated
When customer upgrades to a version  >= 16.0.8e/6.1.1f, Service-ep flag gets removed for the specific service epg vlanCktEp.
This may affect working service graphs.
A warning of failure would help customer look into their L4-L7 config and identify if the PBR config is complete.
TAC should also account for unidirectional PBR config before suggesting any recommendations. 

## How to detect this issue
See attachment
[service-ep flag.txt](https://github.com/user-attachments/files/23337794/service-ep.flag.txt)

### PR DESCRIPTION
**Pull request by jeestr4d**
Fixes #294

Lab output:

[Check 88/88] Service-EP Flag in BD without PBR...                                                                 FAIL - OUTAGE WARNING!!
  Tenant    Bridge Domain        Service Graph Device   Device Node Name
  -------   --------------       --------------------   ----------------
  CIB       Prod-LB-IN_BD        Prod_FTD_LB_Local      N2
  CIB       Prod-LB-Out_BD       Prod-FW-LB-SG          node2
  CIB       Prod-LB-Out_BD       Prod-FW-LB-SG          node2
  CIB       Prod-LB-Out_BD       Prod_FTD_LB_Local      N2
  CIB       Prod-LB-Out_BD       Prod_FTD_LB_Local      N2
  DMZ-TN    BD-L3-MPG01-IWF-EGR  SG-IWF-MPG01-TEMPLATE  IWF-MPG01
  DMZ-TN    BD-L3-MPG02-IWF-EGR  SG-IWF-MPG02-TEMPLATE  IWF-MPG02
  aperezos  PBR_in               FTGA-EW                N1

  Recommended Action:
	Confirm that within these BDs the PBR configuration is complete.
	Please check the reference document for details.
  Reference Document: https://datacenter.github.io/ACI-Pre-Upgrade-Validation-Script/validations/#service-ep-flag-in-bd-without-pbr

jeestrad@JEESTRAD-M-L6VV ACI-Pre-Upgrade-Validation-Script % python3 -m pytest tests/service-ep_flag_bd_check                                                     
============================================================================================= test session starts =============================================================================================
platform darwin -- Python 3.9.6, pytest-8.2.0, pluggy-1.5.0
rootdir: /Users/jeestrad/Documents/Python/pre-upgrade/ACI-Pre-Upgrade-Validation-Script
configfile: pytest.ini
plugins: anyio-4.8.0
collected 7 items                                                                                                                                                                                             

tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs0-5.2(8h)-None-MANUAL CHECK REQUIRED] 
----------------------------------------------------------------------------------------------- live log setup ------------------------------------------------------------------------------------------------
[15:19:43.876 INFO               initialize:5653] Cleaning up previous run files in preupgrade_validator_logs/
[15:19:43.877 INFO               initialize:5655] Creating directories preupgrade_validator_logs/ and preupgrade_validator_logs/json_results/
PASSED                                                                                                                                                                                                  [ 14%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs1-6.0(8h)-6.1(1g)-N/A] PASSED                                                                                  [ 28%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs2-4.2(7s)-5.2(4c)-N/A] PASSED                                                                                  [ 42%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs3-5.2(8h)-6.0(8e)-FAIL - OUTAGE WARNING!!] PASSED                                                              [ 57%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs4-5.2(8h)-6.1(1f)-FAIL - OUTAGE WARNING!!] PASSED                                                              [ 71%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs5-5.2(8h)-6.0(8e)-PASS] PASSED                                                                                 [ 85%]
tests/service-ep_flag_bd_check/test_service-ep_flag_bd_check.py::test_logic[icurl_outputs6-5.2(8h)-6.1(1f)-PASS] PASSED                                                                                 [100%]

============================================================================================== 7 passed in 0.03s ==============================================================================================
jeestrad@JEESTRAD-M-L6VV ACI-Pre-Upgrade-Validation-Script %

### CODE REVIEW

File: aci-preupgrade-validation-script.py [hunk 1]
```diff
@@ -5460,6 +5872,76 @@ def isis_database_byte_check(tversion, **kwargs):
...
+        ) or (
+        # Current version and target version fixed
+            cversion.newer_than("6.0(8e)") and tversion.newer_than("6.0(8e)")
+        ):
+        return Result(result=NA, msg=VER_NOT_AFFECTED)
+
+    bd_dn_regex = r"uni/tn-(?P<bd_tn>[^/]+)/BD-(?P<bd>[^/]+)"
+
+    sg_regex = r"uni/tn-(?P<sg_tn>[^/]+)/"
+    sg_regex += r"ldevCtx-c-(?P<ldev_ctrc>[^-][^g]+)"
```
Comment (monrog2) [inline aci-preupgrade-validation-script.py:5905]: **High — preserve evidence when DN parsing fails.** These character-class expressions reject otherwise valid names containing lowercase `g` in the contract or lowercase `n` in the graph. The code later sets `FAIL_O` even when parsing returns no match, but records neither formatted nor unformatted evidence, yielding a critical result with no affected object. Please use delimiter-aware parsing and append the original `vnsLIfCtx` DN to `unformatted_data` whenever formatting fails, without downgrading the outage status. Add lowercase-name and malformed-DN test cases that assert the returned evidence.

File: aci-preupgrade-validation-script.py [hunk 2]
```diff
@@ -5460,6 +5872,76 @@ def isis_database_byte_check(tversion, **kwargs):
...
+
+    # pbr_regex = r"uni/tn-(?P<pbr_tn>[^/]+)/"
+    # pbr_regex += r"svcCont/svcRedirectPol-(?P<pbr_name>.+)"
+
+    vnsLIfCtx_api = "vnsLIfCtx.json"
+    vnsLIfCtx_api += "?query-target=self&rsp-subtree=children"
+    vnsLIfCtxs = icurl("class", vnsLIfCtx_api)
+
+    for vnsLIfCtx in vnsLIfCtxs:
+        if ("vnsRsLIfCtxToSvcRedirectPol" not in vnsLIfCtx["vnsLIfCtx"]["children"][0]):
```
Comment (monrog2) [inline aci-preupgrade-validation-script.py:5918]: **High — do not depend on APIC child ordering.** This only examines `children[0]`, but APIC child ordering is not a contract. A valid `vnsRsLIfCtxToSvcRedirectPol` later in the list will be missed and produce a false outage warning; an empty list raises `IndexError` and turns the check into `ERROR`. Please search all children for a valid redirect relationship and add shuffled-order, empty-child, and malformed-relation tests.

File: aci-preupgrade-validation-script.py [hunk 3]
```diff
@@ -5460,6 +5872,76 @@ def isis_database_byte_check(tversion, **kwargs):
...
+    doc_url = "https://datacenter.github.io/ACI-Pre-Upgrade-Validation-Script/validations/#service-ep-flag-in-bd-without-pbr"
+
+    if not tversion:
+        return Result(result=MANUAL, msg=TVER_MISSING)
+    if (
+        # Older versions not affected 
+            cversion.older_than("5.2(5c)") and tversion.older_than("5.2(5c)")
+        ) or (
+        # Current version not affected target version fixed
+            cversion.older_than("5.2(5b)") and tversion.newer_than("6.0(8e)")
```
Comment (monrog2) [inline aci-preupgrade-validation-script.py:5895]: **Blocking — use the correct release families and inclusive boundaries.** `tversion` is the target APIC version, but this gate compares it with `6.0(8e)`, while #294 identifies switch `16.0(8e)` / APIC `6.1(1f)` as the fixed boundary. This can classify affected upgrade paths incorrectly, and `older_than` / `newer_than` exclude the exact boundary releases. Please encode the APIC and switch boundaries separately (using `sw_cversion` where required), apply inclusive boundary logic, and add tests immediately below, at, and above `5.2(5c)`, `6.0(1g)`, `16.0(8e)`, and `6.1(1f)`.

File: aci-preupgrade-validation-script.py [hunk 4]
```diff
@@ -5460,6 +5872,76 @@ def isis_database_byte_check(tversion, **kwargs):
...
+def service_ep_flag_bd_check(cversion, tversion, **kwargs):
+    result = PASS
+    headers = ["Tenant ", "Bridge Domain ", "Service Graph Device", "Device Node Name" ]
+    data = []
+    unformatted_headers = ["DN of vnsLIfCtx"]
+    unformatted_data = []
+    recommended_action = (
+        "\n\tConfirm that within these BDs the PBR configuration is complete."
+        "\n\tPlease check the reference document for details."
+    )
```
Comment (monrog2) [inline aci-preupgrade-validation-script.py:5885]: **Medium — direct users to TAC rather than only suggesting a configuration review.** The CSCwi17652 defect evidence and #294 indicate remediation requires Cisco TAC and an internal test API. The current action may encourage unsupported configuration changes or proceeding without remediation. Please explicitly instruct users to contact Cisco TAC before upgrading, and assert the complete `recommended_action` in the tests.

File: admin@10.31.125.151 [hunk 5]
```diff
@@ -0,0 +1,5880 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (monrog2) [inline admin@10.31.125.151:1]: **Medium — remove this accidental repository artifact.** This unrelated 5,880-line copy of the older v3.2 script exposes an internal address in its filename and can confuse packaging and maintenance. Please remove it. The branch should also be rebased onto the current `v4.2.0-dev`; the effective PR currently spans 389 files and GitHub reports it as conflicting.

**Review by monrog2 — CHANGES_REQUESTED** [review]
Requesting changes for the five inline findings:

- Correct APIC/switch release-family gates and inclusive boundaries.
- Make redirect-policy detection independent of child ordering.
- Preserve affected-object evidence when DN parsing fails.
- Direct users to Cisco TAC before upgrade.
- Remove the accidental duplicate script and rebase this conflicting 389-file branch onto current v4.2.0-dev.

Please add the boundary and edge-case tests called out inline.
